src/mscorlib/src/System/Text/EncodingForwarder.cs

```diff
@@ -179,19 +165,169 @@ public unsafe static int GetBytes(Encoding encoding, char[] chars, int charIndex
         public unsafe static int GetBytes(Encoding encoding, char* chars, int charCount, byte* bytes, int byteCount)
         {
             Debug.Assert(encoding != null);
-            if (bytes == null || chars == null)
+            // Validate parameters
+            if ((bytes == null) ||
+                (chars == null) ||
+                (charCount < 0) ||
+                (byteCount < 0))
             {
-                throw new ArgumentNullException(bytes == null ? nameof(bytes) : nameof(chars), Environment.GetResourceString("ArgumentNull_Array"));
-            }
-            if (charCount < 0 || byteCount < 0)
-            {
-                throw new ArgumentOutOfRangeException(charCount < 0 ? nameof(charCount) : nameof(byteCount), Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"));
+                ThrowValidationFailedException(chars, charCount, bytes);
             }
             Contract.EndContractBlock();
 
             return encoding.GetBytes(chars, charCount, bytes, byteCount, encoder: null);
         }
 
+        internal unsafe static bool TryEncode(char* input, int charCount, byte* output, int byteCount, out int charactersConsumed, out int bytesWritten)
+        {
+            const int Shift16Shift24 = (1 << 16) | (1 << 24);
+            const int Shift8Identity = (1 <<  8) | (1);
+
+            int charsToEncode = Math.Min(charCount, byteCount);
+
+            // Encode as bytes upto the first non-ASCII byte and return count encoded
```

Comment:
> 5 utf8 tests in corefx fail; looking into it.

Let me know what you find out. I can take a look as well. It is possible that the corefxlab implementation that I sent the PR for yesterday has bugs because of there is not enough test coverage in corefx for it yet.

> Not sure 100% its the implementation you are after though?

Yes, it is working towards what I have in mind. In addition to that, also delete the redundant implementation of the encoding work horse algorithm in the internal `GetBytes(char* chars, int charCount, ...` method.

The idea is that all GetBytes implementations should follow this pattern:

```
GetBytes(...)
{
    validate arguments

    if there is partial character pending, call method to deal with partial characters/fallbacks/error - this method can be slow and simple

    while (!TryEncode(...))
    {
        call method to deal with partial characters/fallbacks/errors - this method can be slow and simple
    }
}
```

It would allow us to have the exact same code for the fast core encoding algorithm with corefxlab that both I and @KrzysztofCwalina would love to see (https://github.com/dotnet/corefxlab/pull/1156#issuecomment-276141253).

It may make sense to split the change for ASCII and UTF8: Perfect the desired pattern on one first, and then apply it on the other.